interceptor/src/stats/mod.rs

```diff
@@ -10,83 +12,180 @@ pub fn make_stats_interceptor(id: &str) -> Arc<StatsInterceptor> {
     Arc::new(StatsInterceptor::new(id.to_owned()))
 }
 
-#[derive(Debug, Default)]
+#[derive(Debug, Clone)]
+pub struct StreamStats {
+    rtp_recv_stats: RTPStats,
+    rtp_write_stats: RTPStats,
+    rtcp_recv_stats: RTCPStats,
+    rtcp_write_stats: RTCPStats,
+
+    last_update: Instant,
+}
+
+impl Default for StreamStats {
+    fn default() -> Self {
+        Self {
+            rtp_recv_stats: Default::default(),
+            rtp_write_stats: Default::default(),
+            rtcp_recv_stats: Default::default(),
+            rtcp_write_stats: Default::default(),
+            last_update: Instant::now(),
+        }
+    }
+}
+
+impl StreamStats {
+    pub fn snapshot(&self) -> StatsSnapshot {
+        self.into()
+    }
+
+    fn mark_updated(&mut self) {
+        self.last_update = Instant::now();
+    }
+
+    fn duration_since_last_update(&self) -> Duration {
+        self.last_update.elapsed()
+    }
+}
+
+#[derive(Debug)]
+pub struct StatsSnapshot {
+    pub rtp_recv_stats: RTPStats,
+    pub rtp_write_stats: RTPStats,
+    pub rtcp_recv_stats: RTCPStats,
+    pub rtcp_write_stats: RTCPStats,
+}
+
+impl From<&StreamStats> for StatsSnapshot {
+    fn from(stats: &StreamStats) -> Self {
+        Self {
+            rtp_recv_stats: stats.rtp_recv_stats.clone(),
+            rtp_write_stats: stats.rtp_write_stats.clone(),
+            rtcp_recv_stats: stats.rtcp_recv_stats.clone(),
+            rtcp_write_stats: stats.rtcp_write_stats.clone(),
+        }
+    }
+}
+
+#[derive(Default, Debug)]
+struct StatsContainer {
+    stream_stats: HashMap<u32, StreamStats>,
+}
+
+impl StatsContainer {
+    fn get_or_create_stream_stats(&mut self, ssrc: u32) -> &mut StreamStats {
+        self.stream_stats.entry(ssrc).or_default()
+    }
+
+    fn get(&self, ssrc: u32) -> Option<&StreamStats> {
+        self.stream_stats.get(&ssrc)
+    }
+
+    fn remove_stale_entries(&mut self) {
+        self.stream_stats
+            .retain(|_, s| s.duration_since_last_update() < Duration::from_secs(60))
+    }
```

Comment:
Where is this called from?

Reply:
From the stats reducer task. It was the simplest way I could think of to not have unbounded memory growth. It would be better if we knew exactly when a given RTP stream went away so we could clean it up, but that seemed much more complex

Reply:
Ok